server-spi/src/main/java/org/keycloak/models/KeycloakSession.java

```diff
@@ -202,6 +202,8 @@ default <T> T getAttributeOrDefault(String attribute, T defaultValue) {
 
     void close();
 
+    boolean isClosed();
```

Comment:
I am not keen on extending the `KeycloakSession` with this method.

Prevention of throwing an exception is the sole reason for this method. The use of this method exposes `KeycloakSession` internal state. If `KeycloakSession.close` was idempotent and for active transactions handled the transaction commit / rollback itself, this could be completely omitted, and `KeycloakSession`.

Reply:
Sorry, but I'm not following what you mean exposing state. Why exposing if the session is closed so bad?

We are really missing this as well as proper checks to enable a safer usage of the session.

Reply:
The code which handles the transaction is this and is found at many places across the codebase:
```java
        KeycloakTransactionManager tx = session.getTransactionManager();

        try {
            if (tx.isActive()) {
                if (tx.getRollbackOnly()) {
                    tx.rollback();
                } else {
                    tx.commit();
                }
            }
        } finally {
            session.close();
        }
```
There is no place in codebase which closes the session has does not have the transaction commit handling code before itself in try-finally block.

This follows from the fact that closing session without committing the to-be-committed transaction is illegal: The changes in session would be lost, so the session should not exist in the first place.

Hence this code could and should be enclosed in `KeycloakSession.close()` instead of repeating it at many times differently, e.g. [here](https://github.com/keycloak/keycloak/blob/17117820cc14f87f5990ddce80ef38a0e2e7f314/server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java#L255-L277) or [here](https://github.com/keycloak/keycloak/blob/17117820cc14f87f5990ddce80ef38a0e2e7f314/services/src/main/java/org/keycloak/services/filters/AbstractRequestFilter.java#L59-L68). This enclosing also makes `isClosed` method redundant. One of the consequences is that e.g. rollback-only transactions are not handled at all e.g. in [ScheduledTaskRunner](https://github.com/keycloak/keycloak/blob/main/server-spi-private/src/main/java/org/keycloak/services/scheduled/ScheduledTaskRunner.java)

Reply:
I agree with @hmlnarik here. 

It would be much nicer if DefaultKeycloakSession#close handles the closing of the tx itself, so this boilerplate code isn't needed elsewhere.

DefaultKeycloakSession#close could just be a no-op if the session is already closed, or maybe it would be better to throw an exception? Strictly speaking a session that is closed should throw an exception if someone tries to use it.

Reply:
Calling a session `close` twice is illegal operation, so an exception perhaps makes more sense.

> Regarding encapsulating the close/commit, that is the reason why I've added the `TransactionalSessionHandler` to remove boilerplate code between components related to creating/closing _transactional_ sessions within the distribution codebase, without introducing more changes to the core codebase.

I like the way of its separation from the rest of code, so +1 for this.

> Still, I think we should expose whether or not a session is closed. We do have places where a session is only created to obtain providers without any transaction. For instance, setting up background tasks.

Controlling the "any transactions" part by the code is not possible as there is always transaction manager instantiated in a session which is never cleared: https://github.com/keycloak/keycloak/blob/17117820cc14f87f5990ddce80ef38a0e2e7f314/services/src/main/java/org/keycloak/services/DefaultKeycloakSession.java#L94
Furthermore the code which uses the session has no access to actual transactions registered inside TM, so the `commit` / `rollback` operation is performed on TM and never on individual transactions directly, as shown e.g. here: https://github.com/keycloak/keycloak/blob/17117820cc14f87f5990ddce80ef38a0e2e7f314/services/src/main/java/org/keycloak/services/filters/AbstractRequestFilter.java#L64

> IMO, closing a session is purely about releasing providers/resources and not necessarily ending transactions. Keeping these two concepts separate is far more flexible than assuming a session is always transactional.

Should a provider be retrieved on an already closed session, it cannot read or write any data since the data transaction is committed, and - as also proposed in this PR - commit is never performed on an already closed session. Already closed session also contains potentially dirty transaction manager instance since that is never cleared (rightfully so). Thus I don't think that assumption of _closing a session is purely about releasing providers/resources_ is correct. Supporting concept of reusing keycloak session as declared in the current codebase would thus be prone to errors.

Reply:
I see.

I'm not talking about reusing closing sessions, that is what we are discussing avoiding, right? 

My point is more about not assuming all usages of sessions are transactional. And we have this type of usage too.